projects/update_success_nightly_validation.kp/knowledge.md

```diff
@@ -0,0 +1,344 @@
+                                             "payload/targetBuildId",
+                                             "payload/targetChannel",
+                                             "payload/targetVersion",
+                                             "payload/previousBuildId",
+                                             "payload/previousChannel",
+                                             "payload/previousVersion"])
+```
+
+```python
+ping_success = subset.filter(lambda p: p.get("payload/reason") == "success")
+ping_ready = subset.filter(lambda p: p.get("payload/reason") == "ready")
+
+ping_success_count = ping_success.count()
+ping_ready_count = ping_ready.count()
+ping_count = ping_ready_count + ping_success_count
+```
+Quantify the percentage of duplicate pings we're receiving. We don't expect this value to be greater than ~1%, which is the amount we usually get from `main` and `crash`: as a rule of thumb, we threat anything less than 1% as *probably* well behaving.
+
+
+```python
+deduped_subset = dedupe(ping_success, "id")
+deduped_count = deduped_subset.count()
+print("Percentage of duplicate pings: {:.3f}".format(100.0 - pct(deduped_count, ping_success_count)))
+```
+    Percentage of duplicate pings: 0.078
+
+
+The percentage of duplicate pings is within the expected range. Move on and verify the payload of the `update` pings.
+
+
+```python
+def validate_update_payload(p):
+    PAYLOAD_KEYS = [
+        "payload/reason",
+        "payload/previousBuildId",
+        "payload/previousChannel",
+        "payload/previousVersion"
+    ]
+
+    # All the payload keys needs to be strings.
+    for k in PAYLOAD_KEYS:
+        if not isinstance(p.get(k), basestring):
+            return ("'{}' is not a string".format(k), 1)
+        
+    # We only expect "reason" = success.
+    if p.get("payload/reason") != "success":
+        return ("Unexpected 'reason' {}".format(p.get("payload/reason"), 1))
+        
+    # For Nightly, the previous channel should be the same as the
+    # application channel.
+    if p.get("payload/previousChannel") != p.get("application/channel"):
+        return ("Previous channel mismatch: expected {} got {}"\
+                .format(p.get("payload/previousChannel"), p.get("application/channel")), 1)
+                
+    # The previous buildId must be smaller than the application build id.
+    if p.get("payload/previousBuildId") > p.get("application/buildId"):
+        return ("Previous buildId mismatch: {} must be older than {}"\
+                .format(p.get("payload/previoustBuildId"), p.get("application/buildId")), 1)
+    
+    return ("Ok", 1)
+
+validation_results = deduped_subset.map(validate_update_payload).countByKey()
+for k, v in sorted(validation_results.iteritems()):
+    print("{}:\t{:.3f}%".format(k, pct(v, ping_success_count)))
+```
+    Ok:	99.875%
+    Previous buildId mismatch: None must be older than 20170903100443:	0.002%
```

Comment:
These are all errors of the type "`previousBuildId` is `None`", so maybe we should have a specific category for that?

Reply:
Ouch, that was a typo on line 176. Good catch!